File: API/1834_Virtual_Stop_Manager/CS/VirtualStopManagerStrategy.cs
namespace StockSharp.Samples.Strategies;

using System;
using System.Collections.Generic;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

/// <summary>
/// Strategy managing virtual stop loss, take profit, trailing stop and breakeven levels.
/// Converted from MetaTrader script "VR---STEALS-3-EN".
/// </summary>
public class VirtualStopManagerStrategy : Strategy
{
	private readonly StrategyParam<decimal> _volume;
	private readonly StrategyParam<int> _takeProfit;
	private readonly StrategyParam<int> _stopLoss;
	private readonly StrategyParam<int> _trailingStop;
	private readonly StrategyParam<int> _breakeven;
	private readonly StrategyParam<DataType> _candleType;

	/// <summary>
	/// Order volume.
	/// </summary>
	public decimal Volume
	{
		get => _volume.Value;
		set => _volume.Value = value;
	}

	/// <summary>
	/// Virtual take profit in points.
	/// </summary>
	public int TakeProfit
	{
		get => _takeProfit.Value;
		set => _takeProfit.Value = value;
	}

	/// <summary>
	/// Virtual stop loss in points.
	/// </summary>
	public int StopLoss
	{
		get => _stopLoss.Value;
		set => _stopLoss.Value = value;
	}

	/// <summary>
	/// Trailing stop distance in points.
	/// </summary>
	public int TrailingStop
	{
		get => _trailingStop.Value;
		set => _trailingStop.Value = value;
	}

	/// <summary>
	/// Profit in points to move stop to breakeven.
	/// </summary>
	public int Breakeven
	{
		get => _breakeven.Value;
		set => _breakeven.Value = value;
	}

	/// <summary>
	/// Type of candles to process.
	/// </summary>
	public DataType CandleType
	{
		get => _candleType.Value;
		set => _candleType.Value = value;
	}

	private decimal _entryPrice;
	private decimal _stopPrice;
	private decimal _targetPrice;
	private decimal _bestPrice;
	private bool _breakevenMoved;

	/// <summary>
	/// Initializes parameters.
	/// </summary>
	public VirtualStopManagerStrategy()
	{
		_volume = Param(nameof(Volume), 1m)
			.SetGreaterThanZero()
			.SetDisplay("Volume", "Order volume", "General");

		_takeProfit = Param(nameof(TakeProfit), 500)
			.SetGreaterThanZero()
			.SetDisplay("Take Profit (points)", "Virtual take profit in points", "Risk");

		_stopLoss = Param(nameof(StopLoss), 500)
			.SetGreaterThanZero()
			.SetDisplay("Stop Loss (points)", "Virtual stop loss in points", "Risk");

		_trailingStop = Param(nameof(TrailingStop), 300)
			.SetGreaterThanZero()
			.SetDisplay("Trailing Stop (points)", "Trailing stop distance in points", "Risk");

		_breakeven = Param(nameof(Breakeven), 300)
			.SetGreaterThanZero()
			.SetDisplay("Breakeven (points)", "Profit in points to move stop to breakeven", "Risk");

		_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(1).TimeFrame())
			.SetDisplay("Candle Type", "Type of candles to track", "General");
	}

	/// <inheritdoc />
	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
	{
		return new[] { (Security, CandleType) };
	}

	/// <inheritdoc />
	protected override void OnStarted(DateTimeOffset time)
	{
		base.OnStarted(time);

		StartProtection();

		var subscription = SubscribeCandles(CandleType);
		subscription.Bind(ProcessCandle).Start();
	}

	private void ProcessCandle(ICandleMessage candle)
	{
		if (candle.State != CandleStates.Finished)
			return;

		var step = Security.PriceStep ?? 1m;
		var close = candle.ClosePrice;

		// Open long position if none exists.
		if (Position == 0)
		{
			BuyMarket(Volume);
			return;
		}

		if (_entryPrice == 0m)
			return; // Wait for trade price information.

		// Move stop to breakeven when profit exceeds threshold.
		if (!_breakevenMoved && close - _entryPrice >= Breakeven * step)
		{
			_stopPrice = _entryPrice;
			_breakevenMoved = true;
		}

		// Update trailing stop when new high reached.
		if (close > _bestPrice)
		{
			_bestPrice = close;

			if (close - TrailingStop * step > _stopPrice)
				_stopPrice = close - TrailingStop * step;
		}

		// Exit on stop loss or take profit levels.
		if (close <= _stopPrice || close >= _targetPrice)
		{
			SellMarket(Position);
			ResetState();
		}
	}

	/// <inheritdoc />
	protected override void OnNewMyTrade(MyTrade trade)
	{
		base.OnNewMyTrade(trade);

		if (Position == 0 || _entryPrice != 0m)
			return;

		_entryPrice = trade.Trade.Price;
		var step = Security.PriceStep ?? 1m;
		_targetPrice = _entryPrice + TakeProfit * step;
		_stopPrice = _entryPrice - StopLoss * step;
		_bestPrice = _entryPrice;
		_breakevenMoved = false;
	}

	private void ResetState()
	{
		_entryPrice = 0m;
		_stopPrice = 0m;
		_targetPrice = 0m;
		_bestPrice = 0m;
		_breakevenMoved = false;
	}
}

File: API/1834_Virtual_Stop_Manager/PY/virtual_stop_manager_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")
clr.AddReference("StockSharp.Algo.Indicators")
clr.AddReference("StockSharp.Algo.Strategies")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import ExponentialMovingAverage
from StockSharp.Algo.Strategies import Strategy


class virtual_stop_manager_strategy(Strategy):
    def __init__(self):
        super(virtual_stop_manager_strategy, self).__init__()
        self._fast_period = self.Param("FastPeriod", 12) \
            .SetDisplay("Fast Period", "Fast EMA period", "Risk")
        self._slow_period = self.Param("SlowPeriod", 26) \
            .SetDisplay("Slow Period", "Slow EMA period", "Risk")
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromHours(4))) \
            .SetDisplay("Candle Type", "Candle type", "General")
        self._prev_fast = 0.0
        self._prev_slow = 0.0
        self._has_prev = False

    @property
    def fast_period(self):
        return self._fast_period.Value

    @property
    def slow_period(self):
        return self._slow_period.Value

    @property
    def candle_type(self):
        return self._candle_type.Value

    def OnReseted(self):
        super(virtual_stop_manager_strategy, self).OnReseted()
        self._prev_fast = 0.0
        self._prev_slow = 0.0
        self._has_prev = False

    def OnStarted2(self, time):
        super(virtual_stop_manager_strategy, self).OnStarted2(time)
        fast = ExponentialMovingAverage()
        fast.Length = self.fast_period
        slow = ExponentialMovingAverage()
        slow.Length = self.slow_period
        self.SubscribeCandles(self.candle_type) \
            .Bind(fast, slow, self.process_candle) \
            .Start()

    def process_candle(self, candle, fast_val, slow_val):
        if candle.State != CandleStates.Finished:
            return
        fast_val = float(fast_val)
        slow_val = float(slow_val)
        if not self._has_prev:
            self._prev_fast = fast_val
            self._prev_slow = slow_val
            self._has_prev = True
            return
        cross_up = self._prev_fast <= self._prev_slow and fast_val > slow_val
        cross_down = self._prev_fast >= self._prev_slow and fast_val < slow_val
        if cross_up and self.Position <= 0:
            if self.Position < 0:
                self.BuyMarket()
            self.BuyMarket()
        elif cross_down and self.Position >= 0:
            if self.Position > 0:
                self.SellMarket()
            self.SellMarket()
        self._prev_fast = fast_val
        self._prev_slow = slow_val

    def CreateClone(self):
        return virtual_stop_manager_strategy()
